experimental/examples/bubblebench/cursor-shared-tree/src/appState.ts

```diff
@@ -0,0 +1,126 @@
+
+export class AppState implements IAppState {
+    readonly clientsSequenceHelper: SharedTreeSequenceHelper;
+    readonly shortTermSpawnedAnchors: Set<Anchor> = new Set();
+    readonly localClientId: string;
+
+    constructor(
+        private readonly tree: ISharedTree,
+        public width: number,
+        public height: number,
+        numBubbles: number,
+    ) {
+        // Move to root node (which is the Shared AppState Node) and initialize this.clientsSequenceHelper
+        const cursor = tree.forest.allocateCursor();
+        moveToDetachedField(tree.forest, cursor);
+        cursor.enterNode(0);
+        this.clientsSequenceHelper = new SharedTreeSequenceHelper(
+            tree,
+            cursor.buildAnchor(),
+            brand("clients"),
+        );
+        cursor.free();
+
+        // Create the initial JsonableTree for the new local client and insert into the shared tree
+        const initialClientJson = this.makeClientInitialJsonTree(numBubbles);
+        this.localClientId = initialClientJson.clientId;
+        this.clientsSequenceHelper.push(initialClientJson.tree);
+
+        console.log(`created client with id ${this.localClient.clientId} and color ${this.localClient.color}`);
+    }
+
+    public get localClient() {
+        const treeNodes = this.clientsSequenceHelper.getAll();
+        treeNodes.forEach(treeNode => this.shortTermSpawnedAnchors.add(treeNode.anchor));
+
+        const localClientNode = treeNodes
+        .filter(treeNode => treeNode.getFieldValue(Client.clientIdFieldKey) === this.localClientId);
+
+        if (localClientNode.length < 1) {
+           throw new Error('Failed to retreieve local client node');
+        }
+
+        return new Client(this.tree, localClientNode[0].anchor, this.shortTermSpawnedAnchors,);
+    }
+
+    public applyEdits() {
+        this.shortTermSpawnedAnchors.forEach(anchor => {
+            try {
+                this.tree.forest.forgetAnchor(anchor);
+                // console.log('successfully forget anchor')
+            } catch (e) {
+                // console.log("failed to forget anchor.", e)
+            }
+        });
+    }
+
+    makeClientInitialJsonTree(numBubbles: number) {
+        const clientId = `${Math.random()}`;
+        const clientInitialJsonTree: JsonableTree = {
+            type: clientSchema.name,
+            fields: {
+                clientId: [
+                    { type: stringSchema.name, value: clientId },
+                ],
+                color: [{ type: stringSchema.name, value: randomColor() }],
+                bubbles: [],
+            },
+        };
+
+        // create and add initial bubbles to initial client json tree
+        for (let i = 0; i < numBubbles; i++) {
+            const bubble = makeBubble(this.width, this.height);
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            clientInitialJsonTree.fields!.bubbles.push({
```

Comment:
this code would be cleaner (and avoid having to do this not undefined assertion) if you created the array of bubbles before you create the tree that it goes in.

Generally when building tree structures the code comes out cleaner if you build the subtrees first, then build the parent to include them. This avoids having to modify the trees after you construct them.

Reply:
Done in commit 358bf83